src/cargo/core/package_id_spec.rs

```diff
@@ -1,92 +1,30 @@
-/// sufficient to uniquely define a package ID.
-#[derive(Clone, PartialEq, Eq, Debug, Hash, Ord, PartialOrd)]
-pub struct PackageIdSpec {
-    name: String,
-    version: Option<PartialVersion>,
-    url: Option<Url>,
-    kind: Option<SourceKind>,
-}
+pub trait PackageIdSpecQuery {
+    /// Roughly equivalent to `PackageIdSpec::parse(spec)?.query(i)`
+    fn query_str<I>(spec: &str, i: I) -> CargoResult<PackageId>
+    where
+        I: IntoIterator<Item = PackageId>;
 
-impl PackageIdSpec {
-    /// Parses a spec string and returns a `PackageIdSpec` if the string was valid.
-    ///
-    /// # Examples
-    /// Some examples of valid strings
-    ///
-    /// ```
-    /// use cargo::core::PackageIdSpec;
-    ///
-    /// let specs = vec![
-    ///     "https://crates.io/foo",
-    ///     "https://crates.io/foo#1.2.3",
-    ///     "https://crates.io/foo#bar:1.2.3",
-    ///     "https://crates.io/foo#bar@1.2.3",
-    ///     "foo",
-    ///     "foo:1.2.3",
-    ///     "foo@1.2.3",
-    /// ];
-    /// for spec in specs {
-    ///     assert!(PackageIdSpec::parse(spec).is_ok());
-    /// }
-    pub fn parse(spec: &str) -> CargoResult<PackageIdSpec> {
-        if spec.contains("://") {
-            if let Ok(url) = spec.into_url() {
-                return PackageIdSpec::from_url(url);
-            }
-        } else if spec.contains('/') || spec.contains('\\') {
-            let abs = std::env::current_dir().unwrap_or_default().join(spec);
-            if abs.exists() {
-                let maybe_url = Url::from_file_path(abs)
-                    .map_or_else(|_| "a file:// URL".to_string(), |url| url.to_string());
-                bail!(
-                    "package ID specification `{}` looks like a file path, \
-                    maybe try {}",
-                    spec,
-                    maybe_url
-                );
-            }
-        }
-        let mut parts = spec.splitn(2, [':', '@']);
-        let name = parts.next().unwrap();
-        let version = match parts.next() {
-            Some(version) => Some(version.parse::<PartialVersion>()?),
-            None => None,
-        };
-        validate_package_name(name, "pkgid", "")?;
-        Ok(PackageIdSpec {
-            name: String::from(name),
-            version,
-            url: None,
-            kind: None,
-        })
-    }
+    /// Checks whether the given `PackageId` matches the `PackageIdSpec`.
+    fn matches(&self, package_id: PackageId) -> bool;
 
-    /// Roughly equivalent to `PackageIdSpec::parse(spec)?.query(i)`
-    pub fn query_str<I>(spec: &str, i: I) -> CargoResult<PackageId>
+    /// Checks a list of `PackageId`s to find 1 that matches this `PackageIdSpec`. If 0, 2, or
+    /// more are found, then this returns an error.
+    fn query<I>(&self, i: I) -> CargoResult<PackageId>
+    where
+        I: IntoIterator<Item = PackageId>;
+}
+
+impl PackageIdSpecQuery for PackageIdSpec {
```

Comment:
Should we have a simple doc for this? For example

* [`std::os::unix::ffi::OsStrExt`](https://doc.rust-lang.org/stable/std/os/unix/ffi/trait.OsStrExt.html#)
* [`anyhow::Context`](https://docs.rs/anyhow/1.0.75/anyhow/trait.Context.html)
* [`itertools::Itertools`](https://docs.rs/itertools/0.12.0/itertools/trait.Itertools.html) (not exactly an extension trait)